botocore/handlers.py

```diff
@@ -140,6 +143,28 @@ def conditionally_calculate_md5(params, **kwargs):
         calculate_md5(params, **kwargs)
 
 
+def validate_bucket_name(params, **kwargs):
+    # The rules for bucket names in the US Standard region allow bucket names
+    # to be as long as 255 characters, and bucket names can contain any
+    # combination of uppercase letters, lowercase letters, numbers, periods
+    # (.), hyphens (-), and underscores (_).
+    if 'Bucket' not in params:
+        return
+    bucket = params['Bucket']
+    if len(bucket) > 255:
+        error_msg = (
+            'Invalid bucket name "%s": Bucket name cannot be '
+            'longer than 255 characters.' % bucket)
+        raise ParamValidationError(report=error_msg)
+    for char in bucket:
```

Comment:
Using:

```
$ python -m timeit -s 'import botocore.session; from botocore.handlers import validate_bucket_name; params={"Bucket": "my-bucket-name"}' 'validate_bucket_name(params)'
$ python -m timeit -s 'import botocore.session; from botocore.handlers import validate_bucket_name; params={"Bucket": "z9_.-" * 51}' 'validate_bucket_name(params)'
```

I get these results:

| Algo    | Bucket Name    | Time                |
|---------|----------------|---------------------|
| current | my-bucket-name | 1.79 usec per loop  |
| current | 'z9_.-' * 51   | 25.8 usec per loop  |
| set     | my-bucket-name | 1.67 usec per loop  |
| set     | 'z9_.-' * 51   | 16.3 usec per loop  |
| regex   | my-bucket-name | 0.971 usec per loop |
| regex   | 'z9_.-' * 51   | 2.67 usec per loop  |

So not much of a difference between set() and the current approach, but the regex version is faster for the worse case scenario.

I'll switch it over to a regex, and just include the regex in the error msg.